src/validation.cpp

```diff
@@ -3587,15 +3587,15 @@ bool Chainstate::ActivateBestChain(BlockValidationState& state, std::shared_ptr<
         // caused an assert() failure during interrupt in such cases as the UTXO DB flushing checks
         // that the best block hash is non-null.
         if (m_chainman.m_interrupt) break;
+
+        // Write changes periodically to disk, after relay.
+        if (!FlushStateToDisk(state, FlushStateMode::PERIODIC)) {
```

Comment:
Previously we didn't flush before calling `CheckBlockIndex`, but I don't think that would degrade the usefulness of the check. I am not sure though about taking the lock here again. If I read this right the lock is taken three times now on every iteration of the loop with little work being done in between. Might it be useful to consolidate these?

Reply:
> Previously we didn't flush before calling `CheckBlockIndex`

We do flush for every block via `ActivateBestChain` -> `ActivateBestChainStep` -> `ConnectTip` -> `FlushStateToDisk`. However, it is only flushed `IF_NEEDED`, and here we do it `PERIODIC`.

Reply:
> If I read this right the lock is taken three times now on every iteration of the loop with little work being done in between. Might it be useful to consolidate these?

Not a bad idea. How does d2163bd4ccf133bd52f1585e8ef53037afd74df9 look?

Reply:
Actually we can't do that because the lock can't be held for `m_chainman.snapshot_download_completed();` :/.